Durability.md

```diff
@@ -0,0 +1,218 @@
+broadcast mechanism among them. There should still be a direct connection from the
+each safekeeper to the Primary for the acknowledgments though.
+
+Similarly, the responsibility for archiving WAL to S3 can be delegated to one of
+the safekeepers, to reduce the load on the primary.
+
+
+Magic STONITH fairy
+-------------------
+
+Now that we have a system that works as long as only one primary node is running at a time, how
+do we ensure that?
+
+1. Use etcd to grant a lease on a key. The primary node is only allowed to operate as primary
+   when it's holding a valid lease. If the primary node dies, the lease expires after a timeout
+   period, and a new node is allowed to become the primary.
+
+2. Use S3 to store the lease. S3's consistency guarantees are more lenient, so in theory you
+   cannot do this safely. In practice, it would probably be OK if you make the lease times and
+   timeouts long enough. This has the advantage that we don't need to introduce a new
+   component to the architecture.
+
+3. Use Raft or Paxos, with the WAL safekeepers acting as the Acceptors to form the quorum. The
+   next chapter describes this option.
+
+
+Built-in Paxos
+--------------
+
+The WAL safekeepers act as PAXOS Acceptors, and the Processing nodes
+as both Proposers and Learners.
+
+Each WAL safekeeper holds an Epoch value in addition to the VCL and
+the WAL. Each request by the primary to safekeep WAL is accompanied by
+an Epoch value. If a safekeeper receives a request with Epoch that
+doesn't match its current Accepted Epoch, it must ignore (NACK) it.
+(In different Paxos papers, Epochs are called "terms" or "round
+numbers")
+
+When a node wants to become the primary, it generates a new Epoch
+value that is higher than any previously observed Epoch value, and
+globally unique.
+
+
+Accepted Epoch: 555                VCL                   LSN
+                                     |                     |
+                                     V                     V
+.................ccccccccccccccccccccXXXXXXXXXXXXXXXXXXXXXXX
+Archived WAL       Completed WAL          In-flight WAL
+
+
+Primary node startup:
+
+1. Contact all WAL safekeepers that you can reach (if you cannot
+   connect to a quorum of them, you can give up immediately). Find the
+   latest Epoch among them.
+
+2. Generate a new globally unique Epoch, greater than the latest Epoch
+   found in previous step.
+
+2. Send the new Epoch in a Prepare message to a quorum of
+   safekeepers. (PAXOS Prepare message)
+
+3. Each safekeeper responds with a Promise. If a safekeeper has
+   already made a promise with a higher Epoch, it doesn't respond (or
+   responds with a NACK). After making a promise, the safekeeper stops
+   responding to any write requests with earlier Epoch.
+
+4. Once you have received a majority of promises, you know that the
+   VCL cannot advance on the old Epoch anymore. This effectively kills
```

Comment:
Nope. Reellection could happen and old primary could replicate more records of old epoch.

Reply:
After a safekeeper has made a Promise on the new Epoch, it will no longer accept records with the old Epoch. After a majority of safekeepers have made a Promise on the new Epoch, the old primary with old Epoch can no longer get a majority on any changes on the old Epoch, so it can no longer advance the VCL.

If I'm missing something, can you give a step-by-step example of a sequence where that could happen?

Reply:
For simplicity lets have 3 safekeepers A, B, C
Primary P1 writes records for Epoch E1: R1_1, R1_2, R1_3
R1_1 replicated to all A, B, C
R1_2 replicated to B
R1_3 didn't replicated
Primary had got a pause
New primary P2 has promises for epoch E2 from A and C. It designated VCL as R1_1. And it even replicates R2_1 to A. Then it also stalls
Old primary P1 awaken and has promises for epoch E3 from B and C. Now it replicates R1_2 to C and R1_3 to B and C - yep VCL were advanced for epoch E1.

Reply:
Even if considered step 6 (copy WAL from safekeeper with highest LSN), then still VCL will be advanced from R1_1 to R1_2.

Reply:
> For simplicity lets have 3 safekeepers A, B, C
> Primary P1 writes records for Epoch E1: R1_1, R1_2, R1_3 (in format of `R<EPOCH>_<LSN>`)
> R1_1 replicated to all A, B, C
> R1_2 replicated to B
> R1_3 didn't replicated
> Primary had got a pause
> New primary P2 has promises for epoch E2 from A and C. It designated VCL as R1_1. And it even replicates R2_2 to A. Then it also stalls

I'm with you so far.

> Old primary P1 awaken and has promises for epoch E3 from B and C. Now it replicates R1_2 to C and R1_3 to B and C - yep VCL were advanced for epoch E1.

Ah, here's where we differ in terminology, I think. The old primary cannot replicate R1_2 to C; C won't accept it because it already made a promise on E3. Instead, the old primary would replicate R3_2 to C, and R2_3 to B and C. In other word, it would replicate the same piece of WAL with the new Epoch. Does that make sense?

Reply:
> Ah, here's where we differ in terminology, I think. The old primary cannot replicate R1_2 to C; C won't accept it because it already made a promise on E3. Instead, the old primary would replicate R3_2 to C, and R3_3 to B and C. In other words, it would replicate the same piece of WAL with the new Epoch. Does that make sense?

Lets rephrase same thing without "Old Primary".

For simplicity lets have 3 safekeepers A, B, C
Primary P1 writes records for Epoch E1: R1_1(x1), R1_2(x2) (in format of `R<EPOCH>_<LSN>(value)` )
R1_1(x1) replicated to all A, B, C
R1_2(x2) replicated to B
Primary P1 died.

New primary P2 has promises for epoch E2 from A and C. It designated VCL as R1_1. And it wrotes R2_2(y1) to A. Then it also died.

Next primary P3 has got promises for epoch E3 from B and C.
It sees: B has records R1_1(x1) and R1_2(x2), C has record R1_1(x1).
It has to replicate R1_2(x2) to C due to step 6 in this algo.

Which way it will do it? Raft paper says it will be same R1_2(x2).
Do you mean it will be R3_2(x2)?

Reply:
Other example:
P1 performs R1_1(x1), R1_2(x2), R1_3(x3).
A got all R1_1(x1), R1_2(x2) and R1_3(x3).
B got R1_1(x1)
C also got R1_1(x1).
P1 died.

P2 has votes for E2 from B and C. P2 performs R2_2(y1) and wrotes it to B and C. Then it died.

Now we have:
A has R1_1(x1), R1_2(x2) and R1_3(x3)
B has R1_1(x1) and R2_2(y1)
C has R1_1(x1) and R2_2(y1)

Primary P3 has votes for E3 from A and B. Which records it will replicated in which side?

Reply:
Other possible (really) situation:
P1 performs R1_1(x1), R1_2(x2).
A got all R1_1(x1), R1_2(x2).
B got R1_1(x1)
C also got R1_1(x1).
P1 died.

P2 has votes for E2 from B and C. P2 performs R2_2(y1) and wrotes it to B and C. Then it died.

A has R1_1(x1), R1_2(x2)
B has R1_1(x1), R2_2(y1)
C has R1_1(x1).

P3 connects A and C for epoch E3. It replicates R1_2(x2) to C (either as R1_2(x2) or as R3_2(x2). And died.

A has R1_1(x1), R1_2(x2)
B has R1_1(x1), R2_2(y1)
C has R1_1(x1), R1_2(x2) (or R3_2(x2) if you like).

P4 connects to A and B for epoch E4. Which value it should replicate? x2 or y1?

Raft has an answer to this: R1_2(x2) is not considered as committed, because to be considered as committed, P3 have to commit (ie write to both A and C) R3_3(z1).
If it does, then R1_2(x2) is committed, and P4 have to replicate R1_2(x2) and R3_3(z1) to B.
If it doesn't, then P4 have to replicate R2_2(y1) to A. Note, that it still will not be committed until P4 commits R4_3(thau1) as well.

Reply:
Raft is not separate algorithm, as one might thing. Raft is well and thoroughly described MultiPaxos as it should be (ok, with some tuning: placing Proposer, Accepter and Learner at same processes). One have to understand quirks of Raft to clearly understand MultiPaxos.

Reply:
> > Ah, here's where we differ in terminology, I think. The old primary cannot replicate R1_2 to C; C won't accept it because it already made a promise on E3. Instead, the old primary would replicate R3_2 to C, and R3_3 to B and C. In other words, it would replicate the same piece of WAL with the new Epoch. Does that make sense?
> 
> Lets rephrase same thing without "Old Primary".
> 
> For simplicity lets have 3 safekeepers A, B, C
> Primary P1 writes records for Epoch E1: R1_1(x1), R1_2(x2) (in format of `R<EPOCH>_<LSN>(value)` )
> R1_1(x1) replicated to all A, B, C
> R1_2(x2) replicated to B
> Primary P1 died.
> 
> New primary P2 has promises for epoch E2 from A and C. It designated VCL as R1_1. And it wrotes R2_2(y1) to A. Then it also died.
> 
> Next primary P3 has got promises for epoch E3 from B and C.
> It sees: B has records R1_1(x1) and R1_2(x2), C has record R1_1(x1).
> It has to replicate R1_2(x2) to C due to step 6 in this algo.
> 
> Which way it will do it? Raft paper says it will be same R1_2(x2).
> Do you mean it will be R3_2(x2)?

That's right. Kind of. In this scheme, the Epoch number is just a single variable in the safekeeper. The WAL records are not "stamped" with the epoch, the safekeeper only stores its "current" epoch (see also discussion with @arssher above). I should make that more clear in the doc. Rephrasing that scenario with that in mind:

Primary P1 writes records for Epoch E1: R1(x1), R1(x2) (in format of `R<LSN>(value)` ). R1(x1) is replicated to all A, B, C. R2(x2) replicated to only B. Situation now:

A (E1): R1(x1)
B (E1): R1(x1), R2(x2)
C (E1): R1(x1)

Primary P1 dies.

New primary P2 gets promises for epoch E2 from A and C. It designates VCL as R1. And it writes R2(y1) to A. Then it also dies. Situation now:

A (E2): R1(x1), R2(y1)
B (E1): R1(x1), R2(x2)
C (E2): R1(x1)

Next primary P3 gets promises for epoch E3 from B and C.

A (E2): R1(x1), R2(y1)
B (E3): R1(x1), R2(x2)
C (E3): R1(x1)

P3 now replicates R2(x2) to C due to step 6 in this algo:

A (E2): R1(x1), R2(y1)
B (E3): R1(x1), R2(x2)
C (E3): R1(x1), R2(x2)

Reply:
> Other example:
> P1 performs R1_1(x1), R1_2(x2), R1_3(x3).
> A got all R1_1(x1), R1_2(x2) and R1_3(x3).
> B got R1_1(x1)
> C also got R1_1(x1).
> P1 died.
> 
> P2 has votes for E2 from B and C. P2 performs R2_2(y1) and wrotes it to B and C. Then it died.
> 
> Now we have:
> A has R1_1(x1), R1_2(x2) and R1_3(x3)
> B has R1_1(x1) and R2_2(y1)
> C has R1_1(x1) and R2_2(y1)
> 
> Primary P3 has votes for E3 from A and B. Which records it will replicated in which side?

Rephrasing this scenario too so that the safekeeper only stores the current Epoch:

P1 performs R1_1(x1), R1_2(x2), R1_3(x3). A got all R1_1(x1), R1_2(x2) and R1_3(x3). B got R1_1(x1) C also got R1_1(x1). P1 dies. Situation:

A (E1): R1(x1), R2(x2), R3(x3)
B (E1): R1(x1)
C (E1): R1(x1)

P2 has votes for E2 from B and C. P2 performs R2(y1) and writes it to B and C. Then it dies. Now we have:
A (E1): R1(x1), R2(x2), R3(x3)
B (E2): R1(x1), R2(y1)
C (E2): R1(x1), R2(y1)

Primary P3 gets Promises for E3 from A and B:
A (E3): R1(x1), R2(x2), R3(x3)
B (E3): R2(y1), R2(y1)
C (E2): R2(y1), R2(y1)

And I notice that my notation is not sufficient to describe all the state in Paxos. The Acceptors actually store two Epochs: the latest one promised, and the latest one accepted. Rephrasing the above, with the accepted and promised epochs shown separately: 

Primary P3 got Promises for E3 from A and B:
A (a:E1, p:E3): R1(x1), R2(x2), R3(x3)
B (a:E2, p:E3): R2(y1), R2(y1)
C (a:E2, p:E2): R2(y1), R2(y1)

Here "A (a:E1, p:E3)" means that safekeeper has accepted with E1, and promised E3. Now we can resolve the ambiguity. P3 replicates R2(y1) from B to A, because B had a higher accepted epoch than A. Final state is:

A (a:E3, p:E3): R1(x1), R2(y1)
B (a:E2, p:E3): R2(y1), R2(y1)
C (a:E2, p:E2): R2(y1), R2(y1)

Reply:
I'll work through this third scenario too:

> Other possible (really) situation:
> P1 performs R1_1(x1), R1_2(x2).
> A got all R1_1(x1), R1_2(x2).
> B got R1_1(x1)
> C also got R1_1(x1).
> P1 died.
> 
> P2 has votes for E2 from B and C. P2 performs R2_2(y1) and wrotes it to B and C. Then it died.
> 
> A has R1_1(x1), R1_2(x2)
> B has R1_1(x1), R2_2(y1)
> C has R1_1(x1).

The write to C isn't visible here. I think you mean that P2 writes R2_2(y1) to B only. I'll continue with that assumption.

> 
> P3 connects A and C for epoch E3. It replicates R1_2(x2) to C (either as R1_2(x2) or as R3_2(x2). And died.
> 
> A has R1_1(x1), R1_2(x2)
> B has R1_1(x1), R2_2(y1)
> C has R1_1(x1), R1_2(x2) (or R3_2(x2) if you like).
> 
> P4 connects to A and B for epoch E4. Which value it should replicate? x2 or y1?
> 
> Raft has an answer to this: R1_2(x2) is not considered as committed, because to consider it as committed P3 have to commit R3_3(z1) (ie write to both A and C).
> If it does, then R1_2(x2) is committed, and P4 have to replicate R1_2(x2) and R3_3(z1) to B.
> If it doesn't, then P4 have to replicate R2_2(y1) to A. Note, that it still will not be committed until P4 commits R4_3(thau1) as well.

So, rephrasing this scenario from the start with the notation that shows the accepted and promised epochs separately. Starting situation:

A(a:E1, p:E1): R1(x1), R2(x2)
B(a:E1, p:E1): R1(x1)
C(a:E1, p:E1): R1(x1)

P1 dies. P2 gets promises for E2 from B and C. P2 performs R2(y1) and writes it to B. Then it dies, too:

A(a:E1, p:E1): R1(x1), R2(x2)
B(a:E2, p:E2): R1(x1), R2(y1)
C(a:E1, p:E2): R1(x1)

P3 connects to A and C for epoch E3. It replicates R2(x2) to C And dies.

A(a:E1, p:E3): R1(x1), R2(x2)
B(a:E2, p:E2): R1(x1), R2(y1)
C(a:E3, p:E3): R1(x1), R2(x2)

P4 connects to A and B for epoch E4

A(a:E1, p:E4): R1(x1), R2(x2)
B(a:E2, p:E4): R1(x1), R2(y1)
C(a:E3, p:E3): R1(x1), R2(x2)

> Which value it should replicate? x2 or y1?
>
> Raft has an answer to this: R1_2(x2) is not considered as committed, because to consider it as committed P3 have to commit R3_3(z1) (ie write to both A and C).
> If it does, then R1_2(x2) is committed, and P4 have to replicate R1_2(x2) and R3_3(z1) to B.
> If it doesn't, then P4 have to replicate R2_2(y1) to A. Note, that it still will not be committed until P4 commits R4_3(thau1) as well.

Gotcha. Yeah, with Paxos too, R2(x2) is not considered as committed until the accepted value has been updated in a majority of acceptors. So in this case, no value has been committed for R2 yet, and replicating either R2(x2) or R2(y1) would be valid. B has a higher accepted Epoch value, though, so P4 will choose to replicate R2(y1).

A(a:E4, p:E4): R1(x1), R2(y1)
B(a:E2, p:E4): R1(x1), R2(y1)
C(a:E3, p:E3): R1(x1), R2(x2)

R2(y1) still cannot be considered as committed, until B has accepted the new epoch too. So as final step, P4 needs to send an Accept message to B:

A(a:E4, p:E4): R1(x1), R2(y1)
B(a:E4, p:E4): R1(x1), R2(y1)
C(a:E3, p:E3): R1(x1), R2(x2)

Now R2(y1) is committed.

Reply:
> Raft is not separate algorithm, as one might think. Raft is well and thoroughly described MultiPaxos as it should be (ok, with some tuning: placing Proposer, Accepter and Learner at same processes). One have to understand quirks of Raft to clearly understand MultiPaxos.

Sure.

Reply:
> Here "A (a:E1, p:E3)" means that safekeeper has accepted with E1, and promised E3. Now we can resolve the ambiguity. P3 replicates R2(y1) from B to A, because B had a higher accepted epoch than A.

Lets use your terminology.
By the way, then "safekeeper had accepted epoch" happens? Does it happens when WAL record is written? Or it is separate message not bound to WAL?

Lets assume it is bound to WAL record:

P1 performs R1(x1), R1(x2), R1(x3). A got all R1(x1), R1(x2) and R1(x3). B got R1(x1), C also got R1(x1). P1 dies. Situation:

A (E1:E1): R1(x1), R2(x2), R3(x3)
B (E1:E1): R1(x1)
C (E1:E1): R1(x1)

P2 has votes for E2 from B and C. P2 performs R2(y1) and writes it to B. Then it dies. Now we have:
A (E1:E1): R1(x1), R2(x2), R3(x3)
B (E2:E2): R1(x1), R2(y1)
C (E1:E2): R1(x1)

Primary P3 gets Promises for E3 from A and C. It safely replicates R2(x2) to C. Then it dies.
A (E1:E3): R1(x1), R2(x2), R3(x3)
B (E2:E2): R1(x1), R2(y1)
C (E3:E3): R1(x1), R2(x2)

Primary P4 gets promises for E4 from A and B. B had accepted E2, A has accepted E1. Which value P4 should it replicate, x2 or y1?

Now assume "safekeeper had accepted epoch" is a separate process that happens before new (from current safekeeper point of view) WAL records are written.

Well, if proposer (ie primary) waits for acceptance of epoch from majority, then it looks like there no loss of committed records.

For example:
P1 performs R1(x1), R1(x2). A got all R1(x1), R1(x2). B got R1(x1), C also got R1(x1). P1 dies. Situation:

A (E1:E1): R1(x1), R2(x2)
B (E1:E1): R1(x1)
C (E1:E1): R1(x1)

P2 got election (ie promise + accept) from B and C. Then it replicates R2(y1) to B.
A (E1:E1): R1(x1), R2(x2)
B (E2:E2): R1(x1), R2(y1)
C (E2:E2): R1(x1)

Now P3 got election from A and C. Since C has accepted epoch E2, P3 has to truncate log of A to record R1(x1) first:
A (E3:E3): R1(x1)
B (E2:E2): R1(x1), R2(y1)
C (E3:E3): R1(x1)
Then it can write record R2(z1) to A:
A (E3:E3): R1(x1), R2(z1)
B (E2:E2): R1(x1), R2(y1)
C (E3:E3): R1(x1)

Now P4 elected by B and C. Since C has accepted epoch E3, P4 has to truncate B log to R1(x1):
A (E3:E3): R1(x1), R2(z1)
B (E4:E4): R1(x1)
C (E4:E4): R1(x1)

Hey, oops. What if P3 replicated R2(z1) to both A and C ??? Is it committed?
A (E3:E3): R1(x1), R2(z1)
B (E2:E2): R1(x1), R2(y1)
C (E3:E3): R1(x1), R2(z1)

Now P4 elected by B and C and truncates B log (cause it will need to add R2(z1). And then fails.
A (E3:E3): R1(x1), R2(z1)
B (E4:E4): R1(x1)
C (E4:E4): R1(x1), R2(z1)

Next P5 is elected by A and B. Now it have to truncate A log. Therefore R2(z1) became "uncommitted".

Assume we do not truncate log, but overwrite. Then it needs additional record to make mistake. Lets assume after P3 dies we get:
A (E3:E3): R1(x1), R2(z1), R3(z2)
B (E2:E2): R1(x1), R2(y1), R3(y2)
C (E3:E3): R1(x1), R2(z1), R3(z2)

Then P4 elected by B and C. It starts to overwrite B log, but died after overwrite of R2.
A (E3:E3): R1(x1), R2(z1), R3(z2)
B (E4:E4): R1(x1), R2(z1), R3(y2)
C (E4:E4): R1(x1), R2(z1), R3(z2)

Now P5 is elected by A and B. B has higher accepted epoch, therefore P5 replicates its log to A:
A (E5:E5): R1(x1), R2(z1), R3(y2)
B (E5:E5): R1(x1), R2(z1), R3(y2)
C (E4:E4): R1(x1), R2(z1), R3(z2)

Oops, we lost committed z2 value.

Reply:
> Gotcha. Yeah, with Paxos too, R2(x2) is not considered as committed until the accepted value has been updated in a majority of acceptors. So in this case, no value has been committed for R2 yet, and replicating either R2(x2) or R2(y1) would be valid. B has a higher accepted Epoch value, though, so P4 will choose to replicate R2(y1).

Oh, i've missed this comment before I wrote previous message. Yes, with "Paxos doesn't consider ... as committed until ..." it becames correct consensus. Then in my first example in previous message P4 is safe to truncate A log to R1, because R2(x2) is not considered as committed.

Pain point will be to find common log record. Epoch in WAL record really helps to do it. Otherwise some other equality check should be invented.